File: frameworks/node/koa/Implementation/routes/authorized.ts
import type { Context, Middleware, Next } from "koa";

import type { Routes } from "../app.ts";

// rb:wiring authorized.*
/**
 * The token check, as route middleware in front of the handler. ctx.assert throws Koa's 403 for any
 * other token, and Koa's error handling writes it, so the handler never runs.
 */
function requireToken(token: string): Middleware {
  const accepted = `Bearer ${token}`;
  // ctx is annotated because TypeScript accepts a call to an assertion such as ctx.assert only
  // through a name declared with a type.
  return async (ctx: Context, next: Next) => {
    ctx.assert(ctx.get("authorization") === accepted, 403);
    await next();
  };
}
// rb:end

/** authorized: Koa has no authorization of its own, and the Koa organisation publishes none for a bearer token, so the check is route middleware. */
const authorized: Routes = (router, { payloads: p }) => {
  router.get("/authorized/small", requireToken(p.settings.token), (ctx) => {
    ctx.body = p.small;
  });
};

export default authorized;

File: frameworks/go/chi/Implementation/authorized.go
package implementation

import (
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/render"
)

// authorizedRoutes put a middleware in front of the handler that refuses any bearer token but
// settings.json's. chi ships basic authentication and nothing for a bearer token, so the
// middleware is the application's, written as chi's README writes its AdminOnly middleware.
func authorizedRoutes(r chi.Router, p *Payloads) {
	r.With(requireToken(p.Settings.Token)).Get("/authorized/small", func(w http.ResponseWriter, r *http.Request) { render.JSON(w, r, &p.Small) })
}

// rb:wiring authorized.*
func requireToken(token string) func(http.Handler) http.Handler {
	expected := "Bearer " + token
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if r.Header.Get("Authorization") != expected {
				http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}
